psamm/datasource/native.py

```diff
@@ -174,28 +174,44 @@ class NativeModel(object):
     directory subtree that specifies part of the model.
     """
 
-    def __init__(self, path):
-        """Create a model from the specified model file or directory"""
+    def __init__(self, model_from, context=None):
+        """Create a model from the specified model file.
 
+        Model can be a directory, long string, open file, or dictionary.
+        """
+        if isinstance(model_from, string_types):
+            self._model = yaml_load(model_from)
+            self._context = context
+        elif isinstance(model_from, dict):
+            self._context = context
+            self._model = model_from
+        elif hasattr(model_from, 'read') and callable(model_from.read):
+            self._context = context
+            self._model = yaml_load(model_from)
+
+    @classmethod
+    def load_model_from_path(cls, path):
         if os.path.isfile(path):
```

Comment:
It may be better to simply try to open as file (in a `try:` block) then catch the `IOError` if it fails and open as a directory instead. This is almost always better than using `os.path.isfile()` because it avoids a race condition.